### PR TITLE
Implement ContentManager and related services

### PR DESCRIPTION
**Pull request by emmauss**
An attempt in implementing a content management system and services that relies on it. This would serve as a backend for the content services, like ncm, lr, ns and fs-srv. To aid in mounting romfs as paths, I created created a FileSystemProvider interface that would do the actual fs access.
Currently loading user apps are not supported. Requests for systems archives  and loading real fonts from system font archives are implemented.

### CODE REVIEW

**Comment by emmauss** [comment]
Removed some unused stuff and cleaned it up. It should be ready for review.

**Comment by emmauss** [comment]
Currently, the user needs to dump the switch's nand and the copy /nand/system/content/ over to ryufs/nand/system/content/ , while maintaining the directory hierarchy. Only systems archives are supported now. Support for installing from update partitions and user apps can be done after we get a gui.

**Comment by Thog** [comment]
@DrHacknik That doesn't prevent Ryujinx from booting. Like for the shared fonts, Ryujinx will only crash if the game/homebrew try to access resources that aren't present. (e.g If the game requests an archive that the ContentManager doesn't know, Ryujinx will error out telling the user that the archive wasn't found)

**Comment by DrHacknik** [comment]
@Thog That's what I thought. On the other hand, if ContentManager cannot find a resource, and returns an error; then will the emulation continue or will the thread break?

**Comment by Thog** [comment]
The emulation will stop as the comportment would be undefined otherwise.

**Comment by DrHacknik** [comment]
Well that's all I wanted to know. Thanks lol

File: Ryujinx.HLE/HOS/Services/Lr/ILocationResolver.cs [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        private void DeleteContentPath(ServiceCtx Context, long TitleId, ContentType ContentType)
+        {
+            ContentManager ContentManager = Context.Device.System.ContentManager;
+
+            string ContentPath = ContentManager.GetInstalledContentStorage(TitleId, StorageId, ContentType.Manual);
+
+            ContentManager.ClearEntry(TitleId, ContentType.Manual, StorageId);
+        }
+
+        private string ReadUtf8String(ServiceCtx Context, int Index = 0)
```
Comment (Thog) [inline Ryujinx.HLE/HOS/Services/Lr/ILocationResolver.cs]: Identical to the one in IFileSystem, this should be moved to Ryujinx.HLE.Utilities.

File: Ryujinx/Ui/Program.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@
 using Ryujinx.HLE;
 using System;
 using System.IO;
+using System.Linq;
```
Comment (Thog) [inline Ryujinx/Ui/Program.cs]: Probably unused

**Comment by emmauss** [comment]
@Thog addressed comments

File: Ryujinx.HLE/HOS/Services/Lr/ILocationResolver.cs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+            Context.Device.System.ContentManager.RefreshEntries(StorageId, 1);
+
+            return 0;
+        }
+
+        public long ResolveProgramPath(ServiceCtx Context)
+        {
+            long TitleId = Context.RequestData.ReadInt64();
+
+            if(ResolvePath(Context, TitleId, ContentType.Program))
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Lr/ILocationResolver.cs]: `if(` -> `if (`.

File: Ryujinx.HLE/HOS/Services/Ncm/IContentManager.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using Ryujinx.HLE.HOS.Ipc;
+using Ryujinx.HLE.FileSystem;
+
+namespace Ryujinx.HLE.HOS.Services.Ncm
+{
+    class IContentManager :IpcService
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Ncm/IContentManager.cs]: `:IpcService` -> `: IpcService`.

File: Ryujinx.HLE/HOS/Services/Ncm/IContentStorage.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+        public IContentStorage()
+        {
+            m_Commands = new Dictionary<int, ServiceProcessRequest>()
+            {
+                //{8, GetPath }
+            };
+        }
+
+        /*private long GetPath(ServiceCtx Context)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Ncm/IContentStorage.cs]: Why is this commented out?

File: Ryujinx.HLE/HOS/Services/Ncm/IContentStorage.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        private Dictionary<int, ServiceProcessRequest> m_Commands;
+
+        public override IReadOnlyDictionary<int, ServiceProcessRequest> Commands => m_Commands;
+
+        public IContentStorage()
+        {
+            m_Commands = new Dictionary<int, ServiceProcessRequest>()
+            {
+                //{8, GetPath }
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/Ncm/IContentStorage.cs]: Same question. We usually don't keep commented out test code on the codebase.

File: Ryujinx.HLE/HOS/Services/FspSrv/IFileSystemProxy.cs [hunk 7]
```diff
@@ -88,7 +229,34 @@ public void LoadSaveDataFileSystem(ServiceCtx Context)
...
 
-            MakeObject(Context, new IFileSystem(Context.Device.FileSystem.GetGameSavePath(SaveInfo, Context)));
+            string SavePath = Context.Device.FileSystem.GetGameSavePath(SaveInfo, Context);
+
+            FileSystemProvider FileSystemProvider = new FileSystemProvider(SavePath);
+
+            MakeObject(Context, new IFileSystem(SavePath, FileSystemProvider));
+        }
+
+        private string ReadUtf8String(ServiceCtx Context, int Index = 0)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Services/FspSrv/IFileSystemProxy.cs]: You added a `ReadUtf8String` method on `StringUtils`. Why is this one still necessary?

File: Ryujinx.HLE/HOS/Services/FspSrv/FsErr.cs [hunk 8]
```diff
@@ -5,5 +5,6 @@ static class FsErr
         public const int PathDoesNotExist  = 1;
         public const int PathAlreadyExists = 2;
         public const int PathAlreadyInUse  = 7;
+        public const int InvalidInput      = 6001;
```
[the remark this answers is not in the document]
Comment (emmauss) [reply Ryujinx.HLE/HOS/Services/FspSrv/FsErr.cs]: yes, at least according to switchbrew.

**Comment by emmauss** [comment]
addressed comments

**Comment by emmauss** [comment]
rebased and squashed

File: Ryujinx.HLE/HOS/Services/Lr/ILocationResolver.cs [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+            }
+            else
+            {
+                return MakeError(ErrorModule.Lr, LrErr.ProgramLocationEntryNotFound);
+            }
+        }
+
+        private void RedirectPath(ServiceCtx Context, long TitleId, int Flag, ContentType ContentType)
+        {
+            string ContentPath = ReadUtf8String(Context);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Lr/ILocationResolver.cs]: Maybe you could align assignments and casts here?

File: Ryujinx.HLE/HOS/Services/Lr/ILocationResolver.cs [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+
+            LocationEntry NewLocation = new LocationEntry(ContentPath, Flag, TitleId, ContentType);
+
+            Context.Device.System.ContentManager.RedirectLocation(NewLocation, StorageId);
+        }
+
+        private bool ResolvePath(ServiceCtx Context, long TitleId,ContentType ContentType)
+        {
+            ContentManager ContentManager = Context.Device.System.ContentManager;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Lr/ILocationResolver.cs]: Maybe you could align assignments and casts here?

File: Ryujinx.HLE/HOS/Services/Lr/ILocationResolver.cs [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private void DeleteContentPath(ServiceCtx Context, long TitleId, ContentType ContentType)
+        {
+            ContentManager ContentManager = Context.Device.System.ContentManager;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Lr/ILocationResolver.cs]: Maybe you could align assignments and casts here?

File: Ryujinx.HLE/HOS/Services/Set/ISystemSettingsServer.cs [hunk 12]
```diff
@@ -148,5 +165,50 @@ public static long GetSettingsItemValue(ServiceCtx Context)
 
             return 0;
         }
+
+        public static byte[] GetFirmwareData(Switch Device)
+        {
+            byte[] Data = null;
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Set/ISystemSettingsServer.cs]: Maybe you could align assignments and casts here?

File: Ryujinx.HLE/HOS/Services/Set/ISystemSettingsServer.cs [hunk 13]
```diff
@@ -148,5 +165,50 @@ public static long GetSettingsItemValue(ServiceCtx Context)
...
+            long TitleId = 0x0100000000000809;
+
+            string ContentPath = Device.System.ContentManager.GetInstalledContentPath(TitleId, StorageId.NandSystem, ContentType.Data);
+
+            if(string.IsNullOrWhiteSpace(ContentPath))
+            {
+                return null;
+            }
+
+            string FirmwareTitlePath = Device.FileSystem.SwitchPathToSystemPath(ContentPath);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Set/ISystemSettingsServer.cs]: Maybe you could align assignments and casts here?

File: Ryujinx.HLE/HOS/Services/Set/ISystemSettingsServer.cs [hunk 14]
```diff
@@ -148,5 +165,50 @@ public static long GetSettingsItemValue(ServiceCtx Context)
...
+                using (Stream FirmwareFile = FirmwareRomFs.OpenFile("/file"))
+                {
+                    FirmwareFile.CopyTo(MemoryStream);
+                }
+
+                Data = MemoryStream.ToArray();
+            }
+
+            FirmwareContent.Dispose();
+
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/Set/ISystemSettingsServer.cs]: Remove empty line

File: Ryujinx.HLE/FileSystem/Content/ContentManager.cs [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+
+                LocationEntry = NextLocationEntry;
+            }
+        }
+
+        public bool HasNca(string NcaId, StorageId StorageId)
+        {
+            if (ContentDictionary.ContainsValue(NcaId))
+            {
+                var Content = ContentDictionary.FirstOrDefault(x => x.Value == NcaId);
```
[the remark this answers is not in the document]
Comment (emmauss) [reply Ryujinx.HLE/FileSystem/Content/ContentManager.cs]: very little similarities/relation between the instructions here. aligning them won't look nice. same as below

File: Ryujinx.HLE/FileSystem/FileSystemProvider.cs [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+        {
+            return Directory.Exists(Name);
+        }
+
+        public bool FileExists(string Name)
+        {
+            return File.Exists(Name);
+        }
+
+        public long OpenDirectory(string Name, int FilterFlags,out IDirectory DirectoryInterface)
```
Comment (AcK77) [inline Ryujinx.HLE/FileSystem/FileSystemProvider.cs]: Space between args

File: Ryujinx.HLE/FileSystem/FileSystemProvider.cs [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+        {
+            DirectoryInterface = null;
```
Comment (AcK77) [inline Ryujinx.HLE/FileSystem/FileSystemProvider.cs]: Not already null?

File: Ryujinx.HLE/FileSystem/FileSystemProvider.cs [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+
+                return 0;
+            }
+
+            return MakeError(ErrorModule.Fs, FsErr.PathDoesNotExist);
+        }
+
+        public long OpenFile(string Name, out IFile FileInterface)
+        {
+            FileInterface = null;
```
Comment (AcK77) [inline Ryujinx.HLE/FileSystem/FileSystemProvider.cs]: Not already null?

File: Ryujinx.HLE/FileSystem/PFsProvider.cs [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+                return 0;
+            }
+
+            throw new NotSupportedException();
+        }
+
+        public long OpenFile(string Name, out IFile FileInterface)
+        {
+            Name          = Name.TrimStart('/');
```
Comment (AcK77) [inline Ryujinx.HLE/FileSystem/PFsProvider.cs]: Extra spaces

File: Ryujinx.HLE/FileSystem/PFsProvider.cs [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+
+            if (Pfs.FileExists(Name))
+            {
+                Stream Stream = Pfs.OpenFile(Name);
+                FileInterface = new IFile(Stream, Name);
+
+                return 0;
+            }
+
+            FileInterface = null;
```
Comment (AcK77) [inline Ryujinx.HLE/FileSystem/PFsProvider.cs]: Not already null?

File: Ryujinx.HLE/FileSystem/RomFsProvider.cs [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+        }
+
+        public bool FileExists(string Name)
+        {
+            return RomFs.FileExists(Name);
+        }
+
+        public long OpenDirectory(string Name, int FilterFlags, out IDirectory DirectoryInterface)
+        {
+            DirectoryInterface = null;
```
Comment (AcK77) [inline Ryujinx.HLE/FileSystem/RomFsProvider.cs]: Not already null?

File: Ryujinx.HLE/FileSystem/RomFsProvider.cs [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+
+                return 0;
+            }
+
+            return MakeError(ErrorModule.Fs, FsErr.PathDoesNotExist);
+        }
+
+        public long OpenFile(string Name, out IFile FileInterface)
+        {
+            FileInterface = null;
```
Comment (AcK77) [inline Ryujinx.HLE/FileSystem/RomFsProvider.cs]: Not already null?

File: Ryujinx.HLE/HOS/Font/SharedFontManager.cs [hunk 23]
```diff
@@ -32,21 +37,65 @@ public SharedFontManager(Switch Device, long PhysicalAddress)
...
+
+                            uint Start = FontOffset;
+
+                            for (; FontOffset - Start < Data.Length; FontOffset++)
+                            {
+                                Device.Memory.WriteByte(PhysicalAddress + FontOffset, Data[FontOffset - Start]);
+                            }
+
+                            NcaFileStream.Dispose();
+
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Font/SharedFontManager.cs]: Remove empty line

**Pull request merged**